core/src/main/java/com/graphhopper/routing/util/parsers/OSMCarMaxSpeedParser.java

```diff
@@ -0,0 +1,86 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.util.parsers;
+
+import com.graphhopper.reader.ReaderWay;
+import com.graphhopper.routing.profiles.MaxSpeed;
+import com.graphhopper.routing.profiles.DecimalEncodedValue;
+import com.graphhopper.routing.profiles.EncodedValue;
+import com.graphhopper.routing.profiles.EncodedValueLookup;
+import com.graphhopper.routing.util.AbstractFlagEncoder;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.spatialrules.SpatialRule;
+import com.graphhopper.storage.IntsRef;
+import com.graphhopper.util.shapes.GHPoint;
+
+import java.util.List;
+
+import static com.graphhopper.routing.profiles.MaxSpeed.UNSET_SPEED;
+
+public class OSMCarMaxSpeedParser implements TagParser {
+
+    private final DecimalEncodedValue carMaxSpeedEnc;
+    private final double maxPossibleSpeed = MaxSpeed.UNLIMITED_SIGN_SPEED;
+
+    public OSMCarMaxSpeedParser() {
+        this(MaxSpeed.create());
+    }
+
+    public OSMCarMaxSpeedParser(DecimalEncodedValue carMaxSpeedEnc) {
+        this.carMaxSpeedEnc = carMaxSpeedEnc;
+    }
+
+    @Override
+    public void createEncodedValues(EncodedValueLookup lookup, List<EncodedValue> list) {
+        list.add(carMaxSpeedEnc);
+    }
+
+    @Override
+    public IntsRef handleWayTags(IntsRef edgeFlags, ReaderWay way, EncodingManager.Access access, long relationFlags) {
+        double maxSpeed = AbstractFlagEncoder.parseSpeed(way.getTag("maxspeed"));
+
+        if (maxSpeed < 0) {
+            GHPoint estmCentre = way.getTag("estimated_center", null);
+            SpatialRule spatialRule = way.getTag("spatial_rule", null);
+            if (estmCentre != null && spatialRule != null)
+                maxSpeed = spatialRule.getMaxSpeed(way.getTag("highway", ""), maxSpeed);
+        }
+
+        double fwdSpeed = AbstractFlagEncoder.parseSpeed(way.getTag("maxspeed:forward"));
+        if (fwdSpeed < 0 && maxSpeed > 0)
+            fwdSpeed = maxSpeed;
+        if (fwdSpeed > maxPossibleSpeed)
+            fwdSpeed = maxPossibleSpeed;
+
+
+        double bwdSpeed = AbstractFlagEncoder.parseSpeed(way.getTag("maxspeed:backward"));
+        if (bwdSpeed < 0 && maxSpeed > 0)
+            bwdSpeed = maxSpeed;
+        if (bwdSpeed > maxPossibleSpeed)
+            bwdSpeed = maxPossibleSpeed;
+
+        if (fwdSpeed <= 0)
+            fwdSpeed = UNSET_SPEED;
+        carMaxSpeedEnc.setDecimal(false, edgeFlags, fwdSpeed);
+
+        if (bwdSpeed <= 0)
+            bwdSpeed = UNSET_SPEED;
+        carMaxSpeedEnc.setDecimal(true, edgeFlags, bwdSpeed);
```

Comment:
This assumes, that car max speed is stored in 2d. If one might change this in `MaxSpeed`, this might go unnoticed and the speed in the edge will be the backward speed?

This is a bit of an edge case, but I think it would be nice if there would be either a exception or a fallback that uses `Math.min(forward,backward)`. Maybe there is something like this in place already and I missed it :).

Reply:
Yes, there should be probably an exception if not two directions.

Reply:
This is tricky. Or at least we would have to make the twoDirections variable accessible from outside so that this parser can handle the two cases.

Reply:
We could also throw and exception in the encoded value. Something like ``!storeBothDirections && reverse`` in `SimpleIntEncodedValue`. But I am not sure if this would break something elsewhere?